cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```
@@ -71,6 +81,160 @@ public Schema getSchema() {
+   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
+   *                                  is not of {@link Schema.LogicalType#TIMESTAMP_MILLIS} or
+   *                                  {@link Schema.LogicalType#TIMESTAMP_MICROS} type.
+   */
+  public ZonedDateTime getTimestamp(String fieldName, ZoneId zoneId) {
+    Schema.LogicalType logicalType = validateAndGetLogicalType(fieldName, schema.getField(fieldName),
+                                                               new HashSet<>(
+      Arrays.asList(Schema.LogicalType.TIMESTAMP_MILLIS.name(), Schema.LogicalType.TIMESTAMP_MICROS.name())));
+
+    if (fields.get(fieldName) == null) {
+      return null;
+    }
+
+    if (logicalType == Schema.LogicalType.TIMESTAMP_MILLIS) {
+      return getZonedDateTime(fieldName, zoneId, TimeUnit.MILLISECONDS, 1000L);
+    }
+
+    return getZonedDateTime(fieldName, zoneId, TimeUnit.MICROSECONDS, 1000000L);
+  }
+
+  /**
+   * Get zoned date and time represented by the field.
+   *
+   * @param fieldName name of the field
+   * @param zoneId    zone id for the field
+   * @return {@link ZonedDateTime} represented by field.
+   */
+  private ZonedDateTime getZonedDateTime(String fieldName, ZoneId zoneId, TimeUnit unit, long divisor) {
+    Long ts = (Long) fields.get(fieldName);
+    // get fractional portion of timestamp, we always store timestamp in micro seconds, so first get that portion.
+    // for example: timestamp is 1970-01-01T00:00:01.112456Z, here fraction = 112456, tsInSeconds = 1
+    int fraction = (int) (ts % divisor);
+    long tsInSeconds = ts / divisor;
+    // create an Instant with time in seconds and fraction which will be stored as nano seconds.
+    Instant instant = Instant.ofEpochSecond(tsInSeconds, unit.toNanos(fraction));
+    return ZonedDateTime.ofInstant(instant, zoneId);
+  }
+
+  private static Schema.LogicalType validateAndGetLogicalType(String fieldName, Schema.Field field,
```

Why the fieldName and field are passed in separately? That means it requires the caller passing in the correct pair through two arguments, which is not a very good method design. It's better passing in `Schema` and `fieldName`.

---

This is a static method so there is no access to schema object. When a Field is null, we should throw an error message with fieldName for which the error occurred. That is why we have 2 different arguments.

---

Should be `getLogicalType(Schema schema, String fieldName, Set<LogicalType> allowedTypes);`